File: 프로그래머스/lv2/49994. 방문 길이/방문 길이.java
import java.util.*;
class Solution {
    public int solution(String dirs) {
        int answer = 0;
        int x=0;
        int y=0;
        List<String> list=new ArrayList<>();
        for(int i=0;i<dirs.length();i++){
            char c=dirs.charAt(i);
            String pos=Integer.toString(x)+Integer.toString(y);
            switch (c){
                case 'U':
                    y++;
                    break;
                case 'D':
                    y--;
                    break;
                case 'R':
                    x++;
                    break;
                case 'L':
                    x--;
                    break;
                default:
                    break;
            }

            if(x<-5){
                x=-5;
                continue;
            }else if(x>5){
                x=5;
                continue;
            }else if(y<-5){
                y=-5;
                continue;
            }else if(y>5){
                y=5;
                continue;
            }



                String to=Integer.toString(x)+Integer.toString(y);
                String posToGo=pos+to;
                String goToPos=to+pos;

                if(!list.contains(posToGo)&&!list.contains(goToPos)){
                    list.add(posToGo);
                    list.add(goToPos);
                    answer++;
                }




        }

        System.out.println(answer);
        return answer;
    }
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
# def solution(dirs):
#     answer = set() # 중복제거
#     pos = [0,0] # 현재 위치

#     for d in dirs:
#         if d == 'U' and pos[1] < 5:
#             pos[1]+=1
#             now_pos = (pos[0], pos[1] - 0.1)
#         elif d == 'D' and pos[1] > -5:
#             pos[1]+=-1
#             now_pos = (pos[0], pos[1] + 0.1)
#         elif d == 'R' and pos[0] < 5:
#             pos[0]+=1
#             now_pos = (pos[0] - 0.1, pos[1])
#         elif d == 'L' and pos[0] > -5:
#             pos[0]+=-1
#             now_pos = (pos[0] + 0.1, pos[1])
#         else:
#             pass
#         answer.add(now_pos)
#     return len(answer)

def solution(dirs):
    s = set()
    d = {'U': (0,1), 'D': (0, -1), 'R': (1, 0), 'L': (-1, 0)}
    x, y = 0, 0
    for i in dirs:
        nx, ny = x + d[i][0], y + d[i][1]
        if -5 <= nx <= 5 and -5 <= ny <= 5:
            s.add((x,y,nx,ny))
            s.add((nx,ny,x,y))
            x, y = nx, ny
    return len(s)//2